src/ui/sign_tx_nbgl.c

```diff
@@ -0,0 +1,284 @@
+        item_contract_t as_item_contract;
+        item_group_t as_item_group;
+    } content;
+} dynamic_item_t;
+
+typedef struct dynamic_slot_s {
+    char title[64];
+    char text[64];
+} dynamic_slot_t;
+
+static void start_review(void);
+static void rejectUseCaseChoice(void);
+static nbgl_layoutTagValueList_t layout;
+static nbgl_layoutTagValue_t current_pair;
+static nbgl_layoutTagValue_t static_items[MAX_NUM_STEPS + 1];
+static dynamic_slot_t dyn_slots[4];
+static uint8_t static_items_nb;
+static dynamic_item_t dyn_items[64];
+static uint8_t dyn_items_nb;
+
+static int create_transaction_flow(void) {
+    static_items_nb = 0;
+    dyn_items_nb = 0;
+
+    if (!G_context.tx_info.transaction.is_system_asset_transfer && !G_context.tx_info.transaction.is_vote_script &&
+        !N_storage.scriptsAllowed) {
+        return -1;
+    }
+
+    if (G_context.tx_info.transaction.is_vote_script) {
+        if (G_context.tx_info.transaction.is_remove_vote) {
+            static_items[static_items_nb].item = "Object";
+            static_items[static_items_nb].value = "Retracting vote";
+            ++static_items_nb;
+        } else {
+            static_items[static_items_nb].item = "Object";
+            static_items[static_items_nb].value = "Casting vote";
+            ++static_items_nb;
+        }
+    } else if (G_context.tx_info.transaction.is_system_asset_transfer) {
+        static_items[static_items_nb].item = "Object";
+        static_items[static_items_nb].value = "System asset transfer";
+        ++static_items_nb;
+        static_items[static_items_nb].item = "Destination addr";
+        static_items[static_items_nb].value = G_tx.dst_address;
+        ++static_items_nb;
+        static_items[static_items_nb].item = "Token amount";
+        static_items[static_items_nb].value = G_tx.token_amount;
+        ++static_items_nb;
+    }
+
+    static_items[static_items_nb].item = "Target network";
+    static_items[static_items_nb].value = G_tx.network;
+    ++static_items_nb;
+
+    static_items[static_items_nb].item = "System fee";
+    static_items[static_items_nb].value = G_tx.system_fee;
+    ++static_items_nb;
+
+    static_items[static_items_nb].item = "Network fee";
+    static_items[static_items_nb].value = G_tx.network_fee;
+    ++static_items_nb;
+
+    static_items[static_items_nb].item = "Total fees";
+    static_items[static_items_nb].value = G_tx.total_fees;
+    ++static_items_nb;
+
+    static_items[static_items_nb].item = "Valid until height";
+    static_items[static_items_nb].value = G_tx.valid_until_block;
+    ++static_items_nb;
+
+    for (int i = 0; i < G_context.tx_info.transaction.signers_size; ++i) {
+        dyn_items[dyn_items_nb].kind = SIGNER;
+        dyn_items[dyn_items_nb].content.as_item_signer.signer_index = i;
+        ++dyn_items_nb;
+        dyn_items[dyn_items_nb].kind = ACCOUNT;
+        dyn_items[dyn_items_nb].content.as_item_account.signer_index = i;
+        ++dyn_items_nb;
```

Comment:
Couldn't you put some kind of check on `dyn_items_nb` compared to the size of `dyn_items` array ? 

In fantom there's something like that : 

```
static inline void INCR_AND_ASSERT_PAIR_NB(uint8_t * nbPairs) {
    (*nbPairs)++;
    ASSERT(*nbPairs <= MAX_FIELDS_PER_TRANSACTION);
}
```

Reply:
What a raised assertion would do better than a segfault ? There is no way to raise this value to the app as there is no associated APDU... I've bumped the number of items to 256, they are only 3 bytes sized after all so it's still light and it should prevent any issue

Reply:
@fbeutin-ledger true. Maybe if you throw something with an error code (something like incorrect data) it would do a better job ?